### PR TITLE
Shadow module form selection: handle child modules

### PR DESCRIPTION
**Pull request by orangejenny**
http://manage.dimagi.com/default.asp?240834

I got bogged down in the suite.xml changes, particularly how form exclusion works with menu mode, for quite some time, so I wrote tests for all of the variants of menu mode in an app with a parent module, a child module, and a shadow module which has the parent as its source.

@dannyroberts / @kaapstorm

### CODE REVIEW

File: corehq/apps/app_manager/tests/test_shadow_modules.py [hunk 1]
```diff
@@ -12,6 +12,7 @@ def setUp(self):
         self.form1 = self.factory.new_form(self.basic_module)
         self.form1.xmlns = 'http://openrosa.org/formdesigner/secondform'
         self.shadow_module = self.factory.new_shadow_module('shadow_module', self.basic_module, with_form=False)
+        self.child_module, self.form2 = self.factory.new_basic_module('child_module', 'parrot', parent_module=self.basic_module)
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_shadow_modules.py]: E501 line too long (128 > 115 characters)

File: corehq/apps/app_manager/tests/test_shadow_modules.py [hunk 2]
```diff
@@ -66,6 +67,303 @@ def test_no_forms_selected(self):
             "./menu[@id='m1']"
         )
 
+
+    def test_no_child_forms_selected(self):
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_shadow_modules.py]: E303 too many blank lines (2)

File: corehq/apps/app_manager/tests/test_shadow_modules.py (hunk 1 above)
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_shadow_modules.py]: E501 line too long (128 > 115 characters)

File: corehq/apps/app_manager/tests/test_shadow_modules.py (hunk 2 above)
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_shadow_modules.py]: E303 too many blank lines (2)

File: corehq/apps/app_manager/suite_xml/sections/menus.py [hunk 3]
```diff
@@ -123,7 +126,12 @@ def domain_uses_usercase():
                         })
                         menu = Menu(**menu_kwargs)
 
-                    menu.commands.extend(get_commands())
+                    excluded_form_ids = []
+                    if root_module and root_module.doc_type == 'ShadowModule':
```
Comment (dannyroberts) [inline corehq/apps/app_manager/suite_xml/sections/menus.py]: can also do `if isinstance(root_module, ShadowModule):` (unless I'm missing the intention here), which has the advantage of being not just more succinct but would also show up if you did a Find Usages on `ShadowModule` (not just text search).

**Review by dannyroberts — APPROVED** [review]
Great, glad you added tests!

File: corehq/apps/app_manager/suite_xml/sections/menus.py [hunk 4]
```diff
@@ -100,7 +103,7 @@ def domain_uses_usercase():
                     suffix = ""
                     if root_module:
                         menu_kwargs.update({'root': id_strings.menu_id(root_module)})
-                        suffix = id_strings.menu_id(root_module) if root_module.doc_type == 'ShadowModule' else ""
+                        suffix = id_strings.menu_id(root_module) if isinstance(root_module, ShadowModule) else ""
```
Comment (dimagimon) [inline corehq/apps/app_manager/suite_xml/sections/menus.py]: F821 undefined name 'ShadowModule'

File: corehq/apps/app_manager/suite_xml/sections/menus.py [hunk 5]
```diff
@@ -123,7 +126,12 @@ def domain_uses_usercase():
                         })
                         menu = Menu(**menu_kwargs)
 
-                    menu.commands.extend(get_commands())
+                    excluded_form_ids = []
+                    if root_module and isinstance(root_module, ShadowModule):
```
Comment (dimagimon) [inline corehq/apps/app_manager/suite_xml/sections/menus.py]: F821 undefined name 'ShadowModule'

File: corehq/apps/app_manager/suite_xml/sections/menus.py [hunk 6]
```diff
@@ -123,7 +126,12 @@ def domain_uses_usercase():
...
+                        excluded_form_ids = root_module.excluded_form_ids
+                    if id_module and isinstance(id_module, ShadowModule):
```
Comment (dimagimon) [inline corehq/apps/app_manager/suite_xml/sections/menus.py]: F821 undefined name 'ShadowModule'

File: corehq/apps/app_manager/suite_xml/sections/menus.py (hunk 3 above)
Comment (orangejenny) [inline corehq/apps/app_manager/suite_xml/sections/menus.py]: Good point, updated.

**Comment by dannyroberts** [comment]
Sorry my suggestion caused problems. `ImportError: cannot import name ShadowModule`—I think it's just a circular import situation, and you can move the import to inside the using function.

**Comment by orangejenny** [comment]
@dannyroberts Ha, that's what I get for submitting changes without testing them locally. Should be fixed now.

**Pull request merged**